src/main/kotlin/sliv/tool/scene/view/FrameView.kt

```diff
@@ -31,25 +37,51 @@ class FrameView(
     }
 
     fun setFrame(frame: VisualizationFrame) {
-        //TODO: warn if image size doesn't equal to the frame size
+        currentJob?.cancel()
+        landmarksViews = null
+        image = null
+
+        currentJob = scope.launch {
+            reloadData(frame)
+            withContext(Dispatchers.JavaFx) {
+                draw()
+            }
+        }
+        draw()
+    }
+
+    private fun reloadData(frame: VisualizationFrame) {
         landmarksViews = frame.landmarks.mapValues {
             it.value.map { landmark -> LandmarkView.create(landmark) }
         }
-        draw()
+        image = frame.image
+        if(image!!.height != height || image!!.width != width) {
+            println("Image size doesn't equal to the frame size") //TODO: warn user
+        }
     }
 
     private fun draw() {
         val gc = canvas.graphicsContext2D
         gc.clearRect(0.0, 0.0, canvas.width, canvas.height)
-//        gc.drawImage(frame.image, 0.0, 0.0)
-        drawFakeImage() //TODO: real images can be used only with data virtualization
+
+        if(!isDataLoaded) {
+            drawLoadingIndicator()
+            return
+        }
+
+        gc.drawImage(image, 0.0, 0.0, canvas.width, canvas.height)
         doForAllEnabledLandmarks { view -> view.draw(gc, scale.value) }
     }
 
-    private fun drawFakeImage() {
+    private val isDataLoaded
+        get() = landmarksViews != null && image != null
```

Comment:
are these variables assigned in other thread? maybe volatile or thread switch assignment?